File: api-backend/app/schemas/inventory.py
"""
schemas/inventory.py — Pydantic v2 schemas for inventory, products, movements, and suppliers.
"""
from __future__ import annotations
from decimal import Decimal
from typing import List, Optional
from pydantic import BaseModel, Field


# ── Product ─────────────────────────────────────────────────────────
class ProductCreate(BaseModel):
    product_name: str = Field(..., min_length=1, max_length=200)
    category: Optional[str] = None
    subcategory: Optional[str] = None
    brand: Optional[str] = None
    sku: Optional[str] = None
    mrp: Optional[float] = Field(None, ge=0)
    selling_price: Optional[float] = Field(None, ge=0)
    manufacturing_cost: Optional[float] = Field(None, ge=0)
    weight: Optional[float] = Field(None, ge=0)
    fragile_flag: int = 0
    image_url: Optional[str] = None


class ProductUpdate(BaseModel):
    product_name: Optional[str] = Field(None, min_length=1, max_length=200)
    category: Optional[str] = None
    selling_price: Optional[float] = Field(None, ge=0)
    manufacturing_cost: Optional[float] = Field(None, ge=0)
    status: Optional[str] = None
    fragile_flag: Optional[int] = None


class ProductOut(BaseModel):
    id: int
    product_code: Optional[str] = None
    name: str
    category: Optional[str] = None
    sub_category: Optional[str] = None
    brand: Optional[str] = None
    sku: Optional[str] = None
    unit_price: Optional[float] = None
    cost_price: Optional[float] = None
    mrp: Optional[float] = None
    selling_price: Optional[float] = None
    manufacturing_cost: Optional[float] = None
    rating: Optional[float] = None
    review_count: Optional[int] = None
    return_rate: Optional[float] = None
    defect_rate: Optional[float] = None
    description: Optional[str] = None
    weight_kg: Optional[float] = None
    warranty_months: Optional[int] = None
    image_url: Optional[str] = None
    is_fragile: bool = False
    is_active: bool = True
    reorder_point: int = 10
    created_at: Optional[str] = None
    inventory_records: Optional[List['InventoryOut']] = None

    model_config = {"from_attributes": True}


# ── Inventory ────────────────────────────────────────────────────────
class InventoryUpdate(BaseModel):
    current_stock: Optional[int] = Field(None, ge=0)
    reserved_stock: Optional[int] = Field(None, ge=0)
    incoming_stock: Optional[int] = Field(None, ge=0)


class InventoryOut(BaseModel):
    id: int
    product_id: int
    warehouse_id: int
    quantity_on_hand: int = 0
    quantity_reserved: int = 0
    quantity_in_transit: int = 0
    available_quantity: int = 0
    warehouse_location: Optional[str] = None
    warehouse_city: Optional[str] = None
    safety_stock: Optional[int] = None
    reorder_point: Optional[int] = None
    inventory_turnover: Optional[float] = None
    updated_at: Optional[str] = None

    model_config = {"from_attributes": True}


# ── Inventory Movement ───────────────────────────────────────────────
class MovementCreate(BaseModel):
    product_id: int
    movement_type: str = Field(..., pattern="^(STOCK_IN|STOCK_OUT|RETURN|ADJUSTMENT|TRANSFER|DAMAGE)$")
    quantity_delta: int = Field(..., description="Positive=inbound, negative=outbound")
    reference_id: Optional[str] = None
    note: Optional[str] = None


class MovementOut(BaseModel):
    id: int
    product_id: int
    movement_type: str
    quantity_delta: int = 0
    stock_before: int = 0
    stock_after: int = 0
    reference_id: Optional[str] = None
    note: Optional[str] = None
    performed_by: Optional[str] = None
    created_at: Optional[str] = None

    model_config = {"from_attributes": True}


# ── Product-Supplier ─────────────────────────────────────────────────
class ProductSupplierCreate(BaseModel):
    supplier_id: int
    supplier_price: Optional[float] = Field(None, ge=0)
    lead_time_days: Optional[int] = Field(None, ge=0)
    minimum_order_qty: Optional[int] = Field(None, ge=1)
    supplier_rating: Optional[float] = Field(None, ge=0, le=5)
    preferred_supplier_flag: int = 0


class ProductSupplierOut(BaseModel):
    id: int
    product_id: int
    supplier_id: int
    supplier_price: Optional[float] = None
    lead_time_days: Optional[int] = None
    moq: Optional[int] = None
    supplier_rating: Optional[float] = None
    is_preferred: bool = False

    model_config = {"from_attributes": True}


# ── ABC Analysis ─────────────────────────────────────────────────────
class AbcItem(BaseModel):
    product_id: int
    product_name: str
    product_code: Optional[str] = None
    total_revenue: float
    revenue_pct: float
    cumulative_pct: float
    abc_class: str


class AbcAnalysisOut(BaseModel):
    items: List[AbcItem]
    total_products: int
    class_a_count: int
    class_b_count: int
    class_c_count: int


# ── KPI ──────────────────────────────────────────────────────────────
class KpiDefinitionOut(BaseModel):
    id: int
    kpi_code: str
    display_name: str
    description: Optional[str] = None
    formula: Optional[str] = None
    unit: Optional[str] = None
    warning_threshold: Optional[float] = None
    critical_threshold: Optional[float] = None
    higher_is_better: bool = True
    category: Optional[str] = None

    model_config = {"from_attributes": True}

File: web-frontend/src/lib/api/inventory.ts
import { api } from "./client";

// ── Types matching backend schemas ──────────────────────────────────
export interface Product {
  id: number;
  product_code: string | null;
  name: string;
  sku: string;
  category: string | null;
  brand: string | null;
  mrp: number | null;
  selling_price: number | null;
  manufacturing_cost: number | null;
  weight_kg: number | null;
  rating: number | null;
  review_count: number | null;
  warranty_months: number | null;
  return_rate?: number | null;
  defect_rate?: number | null;
  fragile_flag: boolean;
  battery_included: boolean;
  image_url: string | null;
  is_active: boolean;
  inventory_records?: InventoryRecord[];
}

export interface InventoryRecord {
  id: number;
  product_id: number;
  warehouse_id: number;
  quantity_on_hand: number;
  quantity_reserved?: number | null;
  quantity_in_transit?: number | null;
  available_quantity?: number | null;
  safety_stock: number | null;
  reorder_point: number | null;
  warehouse_city: string | null;
  inventory_turnover: number | null;
}

export interface InventoryOut {
  id: number;
  product_id: number;
  warehouse_id: number;
  quantity_on_hand: number;
  quantity_reserved?: number | null;
  quantity_in_transit?: number | null;
  available_quantity?: number | null;
  safety_stock: number | null;
  reorder_point: number | null;
  warehouse_city: string | null;
  inventory_turnover: number | null;
}

export interface AbcItem {
  product_id: number;
  product_name: string;
  product_code: string | null;
  total_revenue: number;
  revenue_pct: number;
  cumulative_pct: number;
  abc_class: "A" | "B" | "C";
}

export interface AbcAnalysisOut {
  items: AbcItem[];
  total_products: number;
  class_a_count: number;
  class_b_count: number;
  class_c_count: number;
}

export interface KpiDefinitionOut {
  id: number;
  kpi_code: string;
  kpi_name: string;
  description: string | null;
  formula: string | null;
  unit: string | null;
  warning_threshold: number | null;
  critical_threshold: number | null;
  higher_is_better: boolean;
  kpi_category: string;
}

export interface MovementOut {
  id: number;
  product_id: number;
  movement_type: string;
  quantity_delta: number;
  stock_before: number;
  stock_after: number;
  performed_by: string | null;
  note: string | null;
  created_at: string;
}

export interface StockAlert {
  product_id: number;
  product_name: string;
  sku: string;
  category: string | null;
  brand: string | null;
  image_url: string | null;
  selling_price: number | null;
  current_stock: number;
  safety_stock: number;
  reorder_point: number;
  warehouse_city: string | null;
  warehouse_id: number;
  incoming_qty: number;
  has_pending_order: boolean;
  pending_order_id: number | null;
  pending_order_qty: number | null;
  pending_order_value?: number | null;
}

// ── API functions ───────────────────────────────────────────────────
export const inventoryApi = {
  listProducts: (params?: {
    category?: string;
    search?: string;
    is_active?: boolean;
    skip?: number;
    limit?: number;
  }) => {
    const q = new URLSearchParams();
    if (params?.category) q.set("category", params.category);
    if (params?.search) q.set("search", params.search);
    if (params?.is_active !== undefined) q.set("is_active", String(params.is_active));
    if (params?.skip !== undefined) q.set("skip", String(params.skip));
    if (params?.limit !== undefined) q.set("limit", String(params.limit));
    return api<Product[]>(`/api/v1/inventory/products?${q}`);
  },

  getProduct: (id: number) => api<Product>(`/api/v1/inventory/products/${id}`),

  getStockAlerts: () => api<StockAlert[]>("/api/v1/inventory/stock-alerts"),

  listStock: (params?: { below_reorder?: boolean; city?: string }) => {
    const q = new URLSearchParams();
    if (params?.below_reorder !== undefined) q.set("below_reorder", String(params.below_reorder));
    if (params?.city) q.set("city", params.city);
    return api<InventoryOut[]>(`/api/v1/inventory/stock?${q}`);
  },

  listMovements: (params?: {
    product_id?: number;
    movement_type?: string;
    skip?: number;
    limit?: number;
  }) => {
    const q = new URLSearchParams();
    if (params?.product_id) q.set("product_id", String(params.product_id));
    if (params?.movement_type) q.set("movement_type", params.movement_type);
    return api<MovementOut[]>(`/api/v1/inventory/movements?${q}`);
  },

  abcAnalysis: () => api<AbcAnalysisOut>("/api/v1/inventory/abc-analysis"),

  listKpis: (category?: string) => {
    const q = category ? `?category=${encodeURIComponent(category)}` : "";
    return api<KpiDefinitionOut[]>(`/api/v1/inventory/kpis${q}`);
  },
};
